numpy/core/numeric.py

```diff
@@ -1544,17 +1542,59 @@ def rollaxis(a, axis, start=0):
     return a.transpose(axes)
 
 
-def _validate_axis(axis, ndim, argname):
+def normalize_axis_tuple(axis, ndim, argname=None, allow_duplicate=False):
```

Comment:
>  whether it would be useful to have this in C

Yes, I think it would - and I think we even already have it for `ufunc.reduce`. I think this is a good stepping stone to implementing that, since it pulls all the python codepaths through one point for parsing axes - that point can now be moved to C later.